### ISSUE TITLE
Drop temporary table doesn't remove relfile

### ISSUE THREAD

**Issue by soumyadeep2007** [greenplum-db/gpdb#7922]
If we create a temp table and then drop it, the relfilenode for the temp table still hangs around on disk on the primaries (it is cleaned up correctly on the master and since we don't WAL temp tables they don't ever hit mirror disks). **So this leaks storage whenever we use temporary tables!**

Now, this is also a problem if say, for example if we have the temp table in a tablespace. We can't drop the tablespace after we attempt to drop the temp table. This is what was encountered here in PR#7896: https://github.com/greenplum-db/gpdb/pull/7896#discussion_r292517175. So we have to fix this test when we fix this issue.

Note: In the repro below t_16384 is the relfilenode for temp table t1.
```
postgres=# create temporary table t1(i int) tablespace ts1;
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
postgres=# drop table t1;
DROP TABLE
postgres=# \q

$ ls -l /Users/pivotal/workspace/gpdb/gpAux/gpdemo/datadirs/*/dem*/base/12810/t_16384
-rw-------  1 pivotal  staff  32768 Jun 12 11:12 /Users/pivotal/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast1/demoDataDir0/base/12810/t_16384
-rw-------  1 pivotal  staff  32768 Jun 12 11:12 /Users/pivotal/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast2/demoDataDir1/base/12810/t_16384
-rw-------  1 pivotal  staff  32768 Jun 12 11:12 /Users/pivotal/workspace/gpdb/gpAux/gpdemo/datadirs/dbfast3/demoDataDir2/base/12810/t_16384
```

### Greenplum version or build
master, 6X_STABLE
### OS version and uname -a
Macos Mojave 10.14.5
Darwin woodlandbay 18.6.0 Darwin Kernel Version 18.6.0: Thu Apr 25 23:16:27 PDT 2019; root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64

**Comment by soumyadeep2007** [comment]
The reason it doesn't remove the relfilenode on the primaries is because we don't persist the `backendId` for `commitrels` in the xlog record in `StartPrepare`. We can see that in `FinishPreparedTransaction`, we read that xlog record and we call `DropRelationFiles`. `DropRelationFiles` further has a bunch of calls to `smgropen` with `InvalidBackendId` and then the subsequent call to`smgrdounlinkall` fails to remove the relfilenodes on disk. This is because `smgrdounlinkall` ultimately calls `mdunlink`->`mdunlinkfork`. Inside that we call `relpath()` -> `GetRelationPath` which returns the path to the relfilenode **without the 't_' prefix**. This is because GetRelationPath assumes that temp tables have a backendId that is NOT `InvalidBackendId`.

**Comment by soumyadeep2007** [comment]
The reason the master can remove the relfilenode correctly is because it calls `smgrdoPendingDeletes()` -> `DropRelationFiles()` with the correct `backendId`.

**Comment by soumyadeep2007** [comment]
`smgrdoPendingDeletes()` is also called on the primary but at the stage it is called the `pendingDeletes` list is empty and thus it is a no-op.

**Comment by soumyadeep2007** [comment]
One way to tackle this problem is to ensure that we persist the backendId for the relations as a part of the struct `RelFileNodeWithStorageType`, such that this information is available to the 2PC mechanism.

**Comment by asimrp** [comment]
Good find and good analysis @soumyadeep2007.  I agree with your proposal to include backend ID along with RelFileNode when storing a relation in prepare WAL record / 2PC state file.

**Comment by asimrp** [comment]
Greenplum only needs a boolean flag to indicate it's a temp relation but if we are to propose this upstream in future, backend ID needs to be stored.

**Comment by soumyadeep2007** [comment]
Closed by #7941 and #7962.
